tests/unit/test_dimension_utilities.py

```diff
@@ -304,6 +320,165 @@ def test_prefetch_dimension_variables(self, mock_get_opendap_nc4):
+
+        # Expected variable contents in file.
+        expected_bounds_data = self.bounds_array
+
+        with self.subTest('Dimension variable is in the root group'):
+            root_variable_full_path = '/npolar_grid_lat'
+            root_varinfo_variable = varinfo_prefetch.get_variable(
+                root_variable_full_path)
+            root_variable_name = 'npolar_grid_lat'
+            root_bounds_name = root_variable_name + '_bnds'
+
+            write_bounds(prefetch_dataset, root_varinfo_variable)
+
+            resulting_bounds_root_data = prefetch_dataset.variables[
+                root_bounds_name][:]
+            assert_array_equal(resulting_bounds_root_data,
+                               expected_bounds_data)
+            # Check that bounds variable exists in the root group.
+            self.assertTrue(prefetch_dataset.variables[root_variable_name])
+            # Check that bounds variable has expected attributes.
+            # (This is only the 'bounds' attribute, since other attributes
+            # are not required.)
+            root_bounds_varinfo_variable = varinfo_prefetch.get_variable(
+                root_variable_full_path)
+            self.assertEqual(root_bounds_varinfo_variable.attributes['bounds'],
+                             root_bounds_name)
+            # Check that dimension variable has 'bounds' attribute.
+            self.assertEqual(root_varinfo_variable.attributes['bounds'],
+                             root_bounds_name)
+            # Check that VariableFromDmr has 'bounds' reference in
+            # the references dictionary.
+            self.assertEqual(root_varinfo_variable.references['bounds'],
+                             {root_bounds_name, })
+
+        with self.subTest('Dimension variable is in a nested group'):
+            nested_variable_full_path = '/group1/group2/zelda'
+            nested_varinfo_variable = varinfo_prefetch.get_variable(
+                nested_variable_full_path)
+            nested_variable_name = 'zelda'
+            nested_variable_path = '/group1/group2'
+            nested_group = prefetch_dataset[nested_variable_path]
+            nested_bounds_name = nested_variable_name + '_bnds'
+
+            write_bounds(prefetch_dataset, nested_varinfo_variable)
+            resulting_bounds_nested_data = nested_group.variables[
+                nested_bounds_name][:]
+            assert_array_equal(resulting_bounds_nested_data,
+                               expected_bounds_data)
+            # Check that bounds variable exists in the nested group.
+            self.assertTrue(nested_group.variables[nested_bounds_name])
+            # Check that bounds variable has expected attributes.
+            # (This is only the 'bounds' attribute, since other attributes
+            # are not required.)
+            nested_bounds_varinfo_variable = varinfo_prefetch.get_variable(
+                nested_variable_full_path)
```

Comment:
I think the same thing applies to this subtest as the previous one:

Looks like you are testing the `bounds` attributes in the `VariableFromDmr` instance for the dimension variable twice, but not checking the `bounds` attribute in the NetCDF-4 file itself. (I _think_???)

Reply:
For #4, I was (falsely, I suppose) thinking that the test

```
# Check that dimension variable has 'bounds' attribute.
self.assertEqual(root_varinfo_variable.attributes['bounds'],
                 root_bounds_name)
```

was referring to the checking the `bounds` attribute of the dimension variable. Am I looking in the wrong place?

Reply:
So that's checking the `VariableFromDmr` instance, which is a good check to do, but I think you also need to check the NetCDF-4 file itself. Something like (based on current Python variable names):

```
self.assertEqual(
    nested_group.variables[nested_variable_name].getncattr('bounds')
    nested_bounds_name
)
```

One thing I don't like about `Variable.getncattr` is that it will raise an exception if there's something missing, so the safer way to do it is probably:

```
self.assertEqual(
    nested_group.variables[nested_variable_name].__dict__.get('bounds')
    nested_bounds_name
)
```

Reply:
I keep forgetting that the variables in the file and the `VariableFromDmr` variables are not the same object. It's kind of obvious though, I suppose.

Reply:
And for #3, the `bounds` doesn't have any attributes because we figured it didn't need any, right? I might be forgetting something though

Reply:
Okay, yeah. Let's skip that one then!